feat: `caml_fatal_error` on misuse of deserialize functions

MisterDA opened the pull request:
Mimicking #12171, this PR gets `intern.c` closer to `extern.c`. It fixes a problem with possible misuse of deserialize functions, as these should only be called from within a `caml_input_*` entrypoint context. Otherwise, the field `intern_src` of `struct caml_intern_state` won't have been populated, and a null pointer dereference occurs.

cc @askvortsov1 and @gasche who authored and reviewed the original PR.

MisterDA committed:
feat: `caml_fatal_error` on misuse of deserialize functions

These should only be called from within a `caml_input_*` entrypoint
context. Otherwise, the field `intern_src` of `struct
caml_intern_state` won't have been populated, and a null pointer
dereference occurs.

kayceesrk commented on Changes:
```diff
@@ -382,8 +382,9 @@ Working version
 
-- #12037: get_extern_state potential NULL dereference.
-  (Alexander Skvortsov, report by Török Edwin,
+- #12037, #12171, #?????: Fix get_extern_state and get_intern_state
```
Best to add a new Changes entry rather than changing the old one. Mixes up attributions unnecessarily.

MisterDA replied:
Right.

kayceesrk commented on runtime/intern.c:
```diff
@@ -224,10 +243,9 @@ static void intern_free_stack(struct caml_intern_state* s)
-    s->intern_stack_limit = s->intern_stack + INTERN_STACK_INIT_SIZE;
   }
+
+  intern_init_stack(s);
```
Why is this moved out of the conditional block? You don't need to `intern_init_stack` if the `intern_stack` is not different from `intern_stack_init`.

MisterDA replied:
For symmetry with `extern.c`. I agree with the comment, so I've also made the change in `extern.c`.

kayceesrk commented on runtime/intern.c:
```diff
@@ -95,8 +95,14 @@ struct caml_intern_state {
+  s->intern_stack_limit = s->intern_stack + INTERN_STACK_INIT_SIZE;
+}
+
+static struct caml_intern_state* prepare_intern_state (void)
```
`init_intern_state` is a better name.

MisterDA replied:
Ok. The name was taken from `extern.c`. For consistency, I've also made the change there.

kayceesrk commented on runtime/intern.c:
```diff
@@ -95,8 +95,14 @@ struct caml_intern_state {
 
-/* Allocates the domain local intern state if needed */
-static struct caml_intern_state* get_intern_state (void)
+static void intern_init_stack(struct caml_intern_state* s)
```
`init_intern_stack` is a better name because we're initializing the `intern_stack`. I don't like the `intern` and `stack` being split in the current name.

MisterDA replied:
Ok. The name was taken from `extern.c`, for consistency I've also made the change there.

kayceesrk commented on runtime/intern.c:
```diff
@@ -110,19 +116,32 @@ static struct caml_intern_state* get_intern_state (void)
+  if (Caml_state->intern_state == NULL)
+    caml_fatal_error (
+      "intern_state not initialized: "
+      "this function can only be called from a `caml_input_*` entrypoint."
```
I'm ok with the first line of the error message "intern_state not initialised". I don't know whether the rest of the message is clarifying the error more "this function can only be called from a `caml_input_*` entrypoint.". 

When will this error occur? Is it that `caml_deserialise_*` was called directly without calling one of the `caml_input_*` functions?

MisterDA replied:
> When will this error occur? Is it that `caml_deserialise_*` was called directly without calling one of the `caml_input_*` functions?

Exactly. In that case, before this PR, the field `intern_src` pointing to the source of the marshaled data won't have been populated, and would be accessed, leading to the nullptr dereference. By splitting initialization of the structure (in `init_intern_state`) and then access (with `get_intern_state`), we can check whether it has been provided a marshaled data source, and error if not (admittedly nicer than a segfault).

kayceesrk replied:
May I request that you add a testcase for this under `testsuite/tests/lib-marshal`? 

Also, the message can be more clear. Instead of

> "this function can only be called from a `caml_input_*` entrypoint."

How about

> "It is likely that caml_deserialize_* function was called without going through `caml_input_*`.

MisterDA replied:
> May I request that you add a testcase for this under testsuite/tests/lib-marshal?

Maybe easier said than done? As the functions raise a fatal error, I can only test one of them by process. A `failwith` would allow to catch the exception and recover.

kayceesrk replied:
> Maybe easier said than done?

This is doable. See `.run` files in the test suite for how it is done. But it is not worth the trouble here as the bytecode and native code behaviours will diverge. Bytecode will perform an intern at 

https://github.com/ocaml/ocaml/blob/225f2f3aebb0540582e61c2ab2b23c57f71d7c73/runtime/startup_byt.c#L566

before user code is run. This will cause any test to not cause a fatal error, which is different from native code.

MisterDA committed:
feat: refactor `intern_free_stack`

Split out `init_extern_stack` into a separate function, so that the
code for initializing and resetting the intern_state stack is not
duplicated.

MisterDA committed:
Make extern.c consistent with changes in intern.c
